### PR TITLE
Adding systemd service to postgresql

### PR DESCRIPTION
**Pull request by KavyaSree2610**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] If you are adding/removing a .spec file that has multiple-versions supported, please add [@microsoft/cbl-mariner-multi-package-reviewers](https://github.com/orgs/microsoft/teams/cbl-mariner-multi-package-reviewers) team as reviewer [(Eg. golang has 2 versions 1.18, 1.21+)](https://github.com/microsoft/azurelinux/tree/2.0/SPECS/golang)
- [x] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?

###### Change Log  

- Added systemd service, installation path, %pre %post and %postun required for the service in the postgresql package

###### Does this affect the toolchain?  

**YES/NO**

###### Associated issues  

- https://microsoft.visualstudio.com/OS/_workitems/edit/49405931

###### Links to CVEs  
- https://nvd.nist.gov/vuln/detail/CVE-YYYY-XXXX

###### Test Methodology

- Pipeline build id: https://dev.azure.com/mariner-org/mariner/_build/results?buildId=610383&view=results

### CODE REVIEW

File: SPECS/postgresql/postgresql.signatures.json [hunk 1]
```diff
@@ -1,5 +1,6 @@
 {
  "Signatures": {
-  "postgresql-16.3.tar.bz2": "331963d5d3dc4caf4216a049fa40b66d6bcb8c730615859411b9518764e60585"
+  "postgresql-16.3.tar.bz2": "331963d5d3dc4caf4216a049fa40b66d6bcb8c730615859411b9518764e60585",
+  "postgresql.service": "918aed2ca0890d0c7894e1f2a5e5c9c20e9b30fdfc1b5a11eef87a873f46271b"
  }
-}
+}
```
Comment (mfrw) [inline SPECS/postgresql/postgresql.signatures.json:6]: NIT: EOF Newline
```suggestion
}

```

File: SPECS/postgresql/postgresql.spec [hunk 2]
```diff
@@ -8,7 +8,7 @@ Distribution:   Azure Linux
 Group:          Applications/Databases
 URL:            https://www.postgresql.org
 Source0:        https://ftp.postgresql.org/pub/source/v%{version}/%{name}-%{version}.tar.bz2
-
+Source1:        %{name}.service
```
Comment (mfrw) [inline SPECS/postgresql/postgresql.spec:11]: We have missed to bump the release number.

File: SPECS/postgresql/postgresql.spec [hunk 3]
```diff
@@ -101,10 +106,59 @@ chown -Rv nobody .
...
+    rm -rf /var/log/%{name}
+    rm -rf /var/run/%{name}
+fi
+
+%systemd_postun_with_restart %{name}.service
 
 %files
 %defattr(-,root,root)
 %license COPYRIGHT
+%{_unitdir}/%{name}.service
```
Comment (mfrw) [inline SPECS/postgresql/postgresql.spec:160]: Let's also add a `changelog` entry

File: SPECS/postgresql/postgresql.spec (hunk 2 above)
Comment (KavyaSree2610) [reply SPECS/postgresql/postgresql.spec:11]: Done

File: SPECS/postgresql/postgresql.spec (hunk 3 above)
Comment (KavyaSree2610) [reply SPECS/postgresql/postgresql.spec:160]: Added

File: SPECS/postgresql/postgresql.signatures.json (hunk 1 above)
Comment (KavyaSree2610) [reply SPECS/postgresql/postgresql.signatures.json]: Done

File: SPECS/postgresql/postgresql.spec [hunk 4]
```diff
@@ -87,6 +90,8 @@ cd contrib && make %{?_smp_mflags}
 %install
 make install DESTDIR=%{buildroot}
 cd contrib && make install DESTDIR=%{buildroot}
+#mkdir -p /usr/local/pgsql/data
```
Comment (mfrw) [inline SPECS/postgresql/postgresql.spec:93]: lets remove this comment.

File: SPECS/postgresql/postgresql.spec [hunk 5]
```diff
@@ -173,6 +227,9 @@ sudo -u nobody -s /bin/bash -c "PATH=$PATH make -k check"
 %{_libdir}/libpgtypes.a
 
 %changelog
+* Wed July 24 Kavya Sree Kaitepalli <kkaitepalli@microsoft.com> - 16.3-2
```
Comment (mfrw) [inline SPECS/postgresql/postgresql.spec:230]:
```suggestion
* Wed Jul 24 Kavya Sree Kaitepalli <kkaitepalli@microsoft.com> - 16.3-2
```

File: SPECS/postgresql/postgresql.spec [hunk 6]
```diff
@@ -1,14 +1,14 @@
 Summary:        PostgreSQL database engine
 Name:           postgresql
 Version:        16.3
-Release:        1%{?dist}
+Release:        2%{?dist}
```
Comment (mfrw) [inline SPECS/postgresql/postgresql.spec:4]: /cc @AZaugg

File: SPECS/postgresql/postgresql.spec (hunk 4 above)
Comment (KavyaSree2610) [reply SPECS/postgresql/postgresql.spec]: Done

File: SPECS/postgresql/postgresql.spec (hunk 5 above)
Comment (KavyaSree2610) [reply SPECS/postgresql/postgresql.spec]: Done

**Comment by KavyaSree2610** [comment]
@microsoft-github-policy-service agree company="Microsoft"

**Comment by KavyaSree2610** [comment]
> @KavyaSree2610 please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
> 1. **Definitions**.
>    **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>    approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>    **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>    Project, including but not limited to communication on electronic mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>    associated comments and documentation.
> 2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>    described in Section 4 below), Submit to any Project.
> 3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work, You may Submit them separately
>    to the Project if You (a) retain all copyright and license information that was in the materials as You
>    received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>    containing materials of a third party:” followed by the names of the third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>    guidelines concerning Submissions.
> 4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your employer has intellectual
>    property rights in Your Submission by contract or applicable law, You must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those Submissions.
> 5. **Licenses**.
> 
> * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>   Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>   the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>   parties.
> * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>   Your patent claims that are necessarily infringed by the Submission or the combination of the
>   Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>   import or otherwise dispose of the Submission alone or with the Project.
> * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>   No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>   granted by implication, exhaustion, estoppel or otherwise.
> 
> 6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>    have disclosed under Section 3). You represent that You have secured permission from Your employer to
>    make the Submission in cases where Your Submission is made in the course of Your work for Your
>    employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>    have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>    You later become aware that would make Your representations in this Agreement inaccurate in any
>    respect.
> 8. **Information about Submissions**. You agree that contributions to Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>    information that You submit with Your Submission.
> 9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
> 10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>     supersedes any and all prior agreements, understandings or communications, written or oral, between
>     the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

File: SPECS/postgresql/postgresql.service [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+After=network-online.target multi-user.target
+Wants=network-online.target
+
+[Service]
+Type=notify
+User=postgres
+ExecStart=/usr/bin/postgres -D /usr/local/pgsql/data
+ExecReload=/bin/kill -HUP $MAINPID
+KillMode=mixed
+KillSignal=SIGINT
```
Comment (AZaugg) [inline SPECS/postgresql/postgresql.service:13]: Is this right?
https://www.postgresql.org/docs/16/server-shutdown.html

I would think Term is better

File: SPECS/postgresql/postgresql.spec [hunk 8]
```diff
@@ -101,10 +105,59 @@ chown -Rv nobody .
...
+
+%post
+
+PGDATA="/usr/local/pgsql/data"
+PGRUN="/run/postgresql"
+
+if [ ! -d "$PGDATA" ]; then
+    mkdir -p "$PGDATA"
+    chown postgres:postgres "$PGDATA"
+    sudo -u postgres /usr/bin/initdb -D "$PGDATA"
```
Comment (AZaugg) [inline SPECS/postgresql/postgresql.spec:126]: I can't recall, but if root is not configured to run sudo commands, ie root ALL(ALL) ALL, won't this fail? 

Would it be better to do
su -c command postgres

File: SPECS/postgresql/postgresql.spec [hunk 9]
```diff
@@ -101,10 +105,59 @@ chown -Rv nobody .
...
+fi
+
+chown -R postgres:postgres "$PGDATA"
```
Comment (AZaugg) [inline SPECS/postgresql/postgresql.spec:129]: As this is done on an installation of a new package, should this be put into the above **if** block? My reasoning is, what if someone already has a mount there with a different UID, could inadvertently change the uid without the user realising.

File: SPECS/postgresql/postgresql.spec [hunk 10]
```diff
@@ -101,10 +105,60 @@ chown -Rv nobody .
...
+    /sbin/useradd -g postgres postgres
+fi
+
+%post
+
+PGDATA="/usr/local/pgsql/data"
+PGRUN="/run/postgresql"
+
+if [ ! -d "$PGDATA" ]; then
+    mkdir -p "$PGDATA"
```
Comment (mfrw) [inline SPECS/postgresql/postgresql.spec:124]: Let's test these changes on a VM.

**Review by mfrw — APPROVED** [review]
LGTM 
Lets test the changes and merge.

**Review by mfrw — APPROVED** [review]
LGTM.
Merging on behalf of @KavyaSree2610 as she does not have merge permissions (yet).

Congratulations on your first PR @KavyaSree2610  :) Welcome to AzureLinux.

**Pull request merged**